builders/shared/default.nix

```diff
@@ -1,17 +1,25 @@
 {
   lib,
-  fetchzip,
+  pkgs,
   runCommand,
 }: {
   generatePubCache = {
-    deps,
+    pubspecNixLock,
     args,
-  }: (runCommand "${args.pname}-pub-cache" {} (lib.mapAttrsToList
-    (path: dep: let
-      derv = fetchzip dep;
-    in ''
-      mkdir -p $out/${dirOf path}
-      ln -s ${derv} $out/${path}
-    '')
-    deps.pub));
+  }: (runCommand "${args.pname}-pub-cache" {} (lib.mapAttrsToList (
+      path: dep: let
+        derv = pkgs.${dep.fetcher} dep.args;
```

Comment:
this is scary to me: i can't think of how exactly it'd be bad but it's untrusted-ish data being fed into nix where there's surely a way to make that bad.

<suggestion supplied out-of-band>

Reply:
done. i used your idea of having a fetchers attrset.